File: android/app/src/main/java/com/naviquest/soundmodule/SoundModule.java
package com.naviquest.soundmodule;

import android.content.Context;
import android.media.MediaPlayer;
import android.media.AudioManager;
import android.net.Uri;
import android.util.Log;
import android.os.Build;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

public class SoundModule extends ReactContextBaseJavaModule {
    private MediaPlayer mediaPlayer;
    private final ReactApplicationContext reactContext;
    
    public SoundModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "SoundModule";
    }

    @ReactMethod
    public void playSound(String soundName, Promise promise) {
        try {
            // Get the resource identifier for the sound file
            Context context = reactContext.getApplicationContext();
            
            Log.d("SoundModule", "=== playSound called ===");
            Log.d("SoundModule", "soundName: " + soundName);
            Log.d("SoundModule", "Package name: " + context.getPackageName());
            
            int resId = context.getResources().getIdentifier(
                    soundName, "raw", context.getPackageName());

            Log.d("SoundModule", "Resource ID lookup result: " + resId);
            
            if (resId == 0) {
                Log.e("SoundModule", "❌ Sound resource NOT found for: " + soundName);
                Log.e("SoundModule", "Available raw resources should include: event_start, checkpoint, event_end, over_speed, time_frame_limit");
                promise.reject("SOUND_NOT_FOUND", "Sound resource not found: " + soundName);
                return;
            }
            
            Log.d("SoundModule", "✅ Found resource ID: " + resId + " for sound: " + soundName);

            // Release any existing media player
            if (mediaPlayer != null) {
                try {
                    mediaPlayer.release();
                    Log.d("SoundModule", "Previous MediaPlayer released");
                } catch (Exception e) {
                    Log.w("SoundModule", "Error releasing previous MediaPlayer: " + e.getMessage());
                }
                mediaPlayer = null;
            }

            // Create and play the new sound with enhanced audio settings
            mediaPlayer = MediaPlayer.create(context, resId);
            if (mediaPlayer == null) {
                Log.e("SoundModule", "❌ Failed to create MediaPlayer for: " + soundName);
                promise.reject("MEDIA_PLAYER_ERROR", "Failed to create MediaPlayer for: " + soundName);
                return;
            }
            
            Log.d("SoundModule", "✅ MediaPlayer created successfully for: " + soundName);
            
            // Set audio attributes for notification sound (better audio focus handling)
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                mediaPlayer.setAudioAttributes(
                        new android.media.AudioAttributes.Builder()
                                .setUsage(android.media.AudioAttributes.USAGE_NOTIFICATION)
                                .setContentType(android.media.AudioAttributes.CONTENT_TYPE_SONIFICATION)
                                .build());
                Log.d("SoundModule", "Audio attributes set");
            }
            
            // Set max volume
            mediaPlayer.setVolume(1.0f, 1.0f);
            Log.d("SoundModule", "Volume set to 1.0");
            
            mediaPlayer.setOnCompletionListener(mp -> {
                Log.d("SoundModule", "✅ Sound playback completed for: " + soundName);
                try {
                    mp.release();
                } catch (Exception e) {
                    Log.w("SoundModule", "Error releasing MediaPlayer on completion: " + e.getMessage());
                }
                mediaPlayer = null;
                promise.resolve("Sound played successfully");
            });
            
            mediaPlayer.setOnErrorListener((mp, what, extra) -> {
                Log.e("SoundModule", "❌ MediaPlayer error - what: " + what + ", extra: " + extra);
                try {
                    mp.release();
                } catch (Exception e) {
                    Log.w("SoundModule", "Error releasing MediaPlayer on error: " + e.getMessage());
                }
                mediaPlayer = null;
                promise.reject("PLAYBACK_ERROR", "Failed to play sound: " + what);
                return true;
            });

            mediaPlayer.start();
            Log.d("SoundModule", "✅ Started playing sound: " + soundName);
            
        } catch (Exception e) {
            Log.e("SoundModule", "❌ Exception in playSound: " + e.getMessage(), e);
            e.printStackTrace();
            promise.reject("SOUND_ERROR", "Error playing sound: " + e.getMessage(), e);
        }
    }

    @ReactMethod
    public void stopSound(Promise promise) {
        // If there's no active mediaPlayer, just resolve immediately
        if (mediaPlayer == null) {
            Log.d("SoundModule", "No active MediaPlayer to stop");
            promise.resolve("No active sound to stop");
            return;
        }
        
        try {
            boolean wasPlaying = false;
            try {
                // Check if it was playing (might throw in some states)
                wasPlaying = mediaPlayer.isPlaying();
                if (wasPlaying) {
                    mediaPlayer.stop();
                    Log.d("SoundModule", "MediaPlayer stopped successfully");
                }
            } catch (IllegalStateException e) {
                // MediaPlayer might be in an invalid state, just log and continue
                Log.w("SoundModule", "MediaPlayer was in an invalid state when checking isPlaying: " + e.getMessage());
                // Don't rethrow, continue with cleanup
            }
            
            try {
                mediaPlayer.release();
                Log.d("SoundModule", "MediaPlayer released successfully");
            } catch (Exception e) {
                // If release fails, just log and continue
                Log.w("SoundModule", "Error releasing MediaPlayer: " + e.getMessage());
                // Don't rethrow, continue with cleanup
            }
            
            // Always nullify the reference
            mediaPlayer = null;
            Log.d("SoundModule", "Sound resources released successfully");
            
            // Always resolve the promise even if there were non-fatal errors
            promise.resolve(wasPlaying ? "Sound stopped and resources released" : "Sound resources released");
            
        } catch (Exception e) {
            // This is a safety catch for any unexpected errors
            // Log error but STILL RESOLVE the promise to avoid unhandled rejections in JS
            Log.e("SoundModule", "Unexpected error in stopSound: " + e.getMessage(), e);
            promise.resolve("Handled unexpected error in stopSound: " + e.getMessage());
        }
    }
    
    @ReactMethod
    public void setVolume(float volume, Promise promise) {
        try {
            if (mediaPlayer != null) {
                // Ensure volume is between 0.0 and 1.0
                float safeVolume = Math.max(0.0f, Math.min(1.0f, volume));
                mediaPlayer.setVolume(safeVolume, safeVolume);
                Log.d("SoundModule", "Volume set to: " + safeVolume);
                promise.resolve("Volume set successfully");
            } else {
                Log.w("SoundModule", "No active media player to set volume");
                promise.resolve("No active media player to set volume");
            }
        } catch (Exception e) {
            Log.e("SoundModule", "Error setting volume: " + e.getMessage(), e);
            promise.resolve("Error setting volume - resolved safely: " + e.getMessage());
        }
    }
}

File: ios/NaviQuest/SoundModule.swift
import Foundation
import AVFoundation
import React

@objc(SoundModule)
class SoundModule: NSObject {
  
  private var audioPlayer: AVAudioPlayer?
  private var currentPromiseResolve: RCTPromiseResolveBlock?
  private var currentPromiseReject: RCTPromiseRejectBlock?
  
  override init() {
    super.init()
    // Configure audio session for playback
    #if os(iOS)
    do {
      try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default, options: [.mixWithOthers, .duckOthers])
      try AVAudioSession.sharedInstance().setActive(true)
      print("SoundModule: Audio session configured successfully")
    } catch {
      print("SoundModule: Failed to configure audio session: \(error.localizedDescription)")
    }
    #endif
  }
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  @objc
  func playSound(_ soundName: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    print("SoundModule: === playSound called ===")
    print("SoundModule: soundName: \(soundName)")
    
    // Stop any currently playing sound
    if audioPlayer != nil {
      audioPlayer?.stop()
      audioPlayer = nil
      print("SoundModule: Previous audio player stopped")
    }
    
    // Store promise callbacks
    currentPromiseResolve = resolve
    currentPromiseReject = reject
    
    // Try to find the sound file in the bundle
    // First try main bundle
    var soundURL: URL?
    
    // Try to find the mp3 file in the main bundle
    if let url = Bundle.main.url(forResource: soundName, withExtension: "mp3") {
      soundURL = url
      print("SoundModule: ✅ Found sound in main bundle: \(url.path)")
    } else if let url = Bundle.main.url(forResource: soundName, withExtension: "wav") {
      soundURL = url
      print("SoundModule: ✅ Found sound in main bundle (wav): \(url.path)")
    } else if let url = Bundle.main.url(forResource: soundName, withExtension: "m4a") {
      soundURL = url
      print("SoundModule: ✅ Found sound in main bundle (m4a): \(url.path)")
    }
    
    // If not found, try searching in Sounds folder
    if soundURL == nil {
      if let url = Bundle.main.url(forResource: soundName, withExtension: "mp3", subdirectory: "Sounds") {
        soundURL = url
        print("SoundModule: ✅ Found sound in Sounds folder: \(url.path)")
      }
    }
    
    guard let url = soundURL else {
      print("SoundModule: ❌ Sound resource NOT found for: \(soundName)")
      print("SoundModule: Available sounds should include: event_start, checkpoint, event_end, over_speed, time_frame_limit")
      let error = NSError(domain: "SoundModule", code: 404, userInfo: [NSLocalizedDescriptionKey: "Sound resource not found: \(soundName)"])
      reject("SOUND_NOT_FOUND", "Sound resource not found: \(soundName)", error)
      return
    }
    
    do {
      // Ensure audio session is active on iOS
      #if os(iOS)
      try AVAudioSession.sharedInstance().setActive(true)
      #endif
      
      audioPlayer = try AVAudioPlayer(contentsOf: url)
      audioPlayer?.delegate = self
      audioPlayer?.volume = 1.0
      audioPlayer?.prepareToPlay()
      
      if audioPlayer?.play() == true {
        print("SoundModule: ✅ Started playing sound: \(soundName)")
      } else {
        print("SoundModule: ❌ Failed to start playback for: \(soundName)")
        let error = NSError(domain: "SoundModule", code: 500, userInfo: [NSLocalizedDescriptionKey: "Failed to start playback"])
        reject("PLAYBACK_ERROR", "Failed to start playback for: \(soundName)", error)
      }
    } catch {
      print("SoundModule: ❌ Exception creating audio player: \(error.localizedDescription)")
      reject("SOUND_ERROR", "Error playing sound: \(error.localizedDescription)", error)
    }
  }
  
  @objc
  func stopSound(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    print("SoundModule: stopSound called")
    
    if audioPlayer == nil {
      print("SoundModule: No active audio player to stop")
      resolve("No active sound to stop")
      return
    }
    
    audioPlayer?.stop()
    audioPlayer = nil
    currentPromiseResolve = nil
    currentPromiseReject = nil
    print("SoundModule: Audio player stopped successfully")
    resolve("Sound stopped successfully")
  }
  
  @objc
  func setVolume(_ volume: Float, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    print("SoundModule: setVolume called with volume: \(volume)")
    
    let safeVolume = max(0.0, min(1.0, volume))
    audioPlayer?.volume = safeVolume
    resolve("Volume set to \(safeVolume)")
  }
}

// MARK: - AVAudioPlayerDelegate
extension SoundModule: AVAudioPlayerDelegate {
  func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
    print("SoundModule: ✅ Sound playback completed, success: \(flag)")
    
    if let resolve = currentPromiseResolve {
      resolve("Sound played successfully")
    }
    
    currentPromiseResolve = nil
    currentPromiseReject = nil
    audioPlayer = nil
  }
  
  func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error: Error?) {
    print("SoundModule: ❌ Audio player decode error: \(error?.localizedDescription ?? "Unknown error")")
    
    if let reject = currentPromiseReject {
      let nsError = error as NSError? ?? NSError(domain: "SoundModule", code: 500, userInfo: [NSLocalizedDescriptionKey: "Audio decode error"])
      reject("DECODE_ERROR", "Audio decode error: \(error?.localizedDescription ?? "Unknown")", nsError)
    }
    
    currentPromiseResolve = nil
    currentPromiseReject = nil
    audioPlayer = nil
  }
}
